### PR TITLE
[#1108] Check that spritesheet is valid given source texture dimensions

### LINKED ISSUE #1108: [chore] Check that spritesheet is valid given source texture dimensions
### Context
Was a TODO #1105 

It is currently possible to create a spritesheet that is invalid given it's source texture dimensions
See https://github.com/excaliburjs/Excalibur/pull/1106/files#diff-c3d2f62c7d54f809d5dc0e29d6e45f7dL62

### Proposal
* [ ] Validate spritesheet against the texture's loaded dimensions.
* [ ] Write a test that checks this validation works

### PR DESCRIPTION
**Pull request by edualb**
===:clipboard: PR Checklist :clipboard:===
- [x] :pushpin: issue exists in github for these changes 
- [x] :microscope: existing tests still pass
- [x] :triangular_ruler: new tests written and passing / old tests updated with new scenario(s)
- [x] :page_facing_up: changelog entry added (or not needed)

==================

Closes #1108 

## Changes:

- It was updated the package.json and package-lock.json with current versions
- It was implemented two exceptions in SpriteSheet in which check if the image Texture size is correct.
- It was implemented two tests for check if the exceptions is correct.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -56,20 +56,20 @@
...
     "jasmine": "2.99.0",
     "jasmine-core": "2.99.1",
-    "karma": "3.1.0",
+    "karma": "^4.0.1",
     "karma-chrome-launcher": "2.2.0",
     "karma-coverage": "1.1.2",
     "karma-coverage-istanbul-reporter": "^2.0.4",
     "karma-jasmine": "1.1.0",
-    "karma-typescript": "3.0.13",
+    "karma-typescript": "^4.0.0",
```
Comment (eonarheim) [inline package.json]: Please revert `package.json` & `package-lock.json`

**Comment by edualb** [comment]
I will check this errors...

**Comment by edualb** [comment]
@eonarheim So, there is something very wrong with the tests. When I run `npm run all`, all tests are successfully. When I run `npm run test` one more time, I'm received differents results,some instants with succesfully and other instants with errors. Do you have any idea to solved it? I'll commit my changes to you check.

**Comment by edualb** [comment]
Travis-CI got one of errors.
This is that showing in my computer:

image

If you want I can make a video to show you.

**Comment by eonarheim** [comment]
@edualb Odd, locally I get a different number of errors. I wonder if we are running into a platform specific bug. TravisCI is running linux, my machine is windows, what's yours?

image

**Comment by edualb** [comment]
@eonarheim windows 10. I receive this error too. If you run many times in a some moment all tests are successfully. Really, is odd.

**Comment by eonarheim** [comment]
@edualb Poking around I think the issue is that the backing texture may or may not be loaded at constructor time. Adding some defense and checking if the naturalWidth exists and is non-zero seems to fix the flaky tests. This can only happen if the texture is not yet loaded.

image

```typescript
if (this.image instanceof Texture && 
        (this.image && this.image.image)) {
      if (this.image.image.naturalWidth && 
          this.spWidth * this.columns > this.image.image.naturalWidth) {
        throw new Error(`SpriteSheet specified is wider, 
        ${this.columns} cols x ${this.spWidth} pixels > ${this.image.image.naturalWidth} 
        pixels than image width`);
      }
      if (this.image.image.naturalHeight && 
          this.spHeight * this.rows > this.image.image.naturalHeight) {
        throw new Error(`SpriteSheet specified is taller, 
        ${this.rows} rows x ${this.spHeight} pixels > ${this.image.image.naturalHeight} 
        pixels than image height`);
      }
    }
```
Since we do not force people to only use loaded textures, instead of checking at constructor time, we should probably wait until the texture is loaded something like 

```typescript
if(this.image) {
  this.image.loaded.then(() => {
    // check image height/width and throw
 });
}
```

**Comment by edualb** [comment]
@eonarheim 
So, I think put this code inside of constructor will work:
```
    if (this.image instanceof Texture) {
        this.image.load().then((imageAux) => {
            if (this.spWidth * this.columns > imageAux.naturalWidth) {
                throw new Error(`SpriteSheet specified is wider, 
                    ${this.columns} cols x ${this.spWidth} pixels > ${this.image.image.naturalWidth} 
                    pixels than image width`);
            }
            if (this.spHeight * this.rows > imageAux.naturalHeight) {
                throw new Error(`SpriteSheet specified is taller,
                    ${this.rows} rows x ${this.spHeight} pixels > ${this.image.image.naturalHeight} 
                    pixels than image height`);
            }
        });
    }
```
With this code when the image is loaded we will can check the throws errors.

But when I compile i receive this error:

`Uncaught Error: Cannot resolve a promise that is not in a pending state! thrown`

Do you have any idea what is wrong with my code?

**Comment by eonarheim** [comment]
@edualb Switch `.load()` to `.loaded`, I think the issue is because `.load()` is used by the excalibur load machinery

**Comment by edualb** [comment]
@eonarheim I changed `.load` to `.loaded`. Yesterday I was trying run tests with this modifications but i think exists some "delay" between resolve promise and finish constructor. Because when I put a `console.log` before conditions to throw error, in the tests shows this `console.log` but dont shows the throw error. Can you help me solved this test? I will commit my modifications and you can check it.

**Comment by edualb** [comment]
@eonarheim Do you know any way to  put a catch block in .loaded Promise? Because when I try put catch block after then block I receive this error: Property 'catch' does not exist on type 'Promise<any>'.

Something like this:
```
        this.image.loaded.then((image: HTMLImageElement) => {
            try {
                if (this.spWidth * this.columns > image.naturalWidth) {
                    throw new RangeError(`SpriteSheet specified is wider, 
                        ${this.columns} cols x ${this.spWidth} pixels > ${image.naturalWidth} 
                        pixels than image width`);
                }
                if (this.spHeight * this.rows > image.naturalHeight) {
                    throw new RangeError(`SpriteSheet specified is taller,
                        ${this.rows} rows x ${this.spHeight} pixels > ${image.naturalHeight} 
                        pixels than image height`);
                }
            } catch (e) {
                return e;
            }
        }).catch(e => {
            return e;
        });
```

**Comment by eonarheim** [comment]
@edualb Unfortunately Excalibur's promise implementation is incomplete (it predates the ES promise specification) it is on our roadmap to remove and replace with native browser promises #994 (@alanag13 has expressed interest in fixing this)

Instead of re-throwing, for now lets use the ex.Logger to log and error

**Comment by edualb** [comment]
@eonarheim  this way is good to merge? 

More later i'll check the #994 =)

**Pull request merged**